Sources/X509/Verifier/ServerIdentityPolicy.swift

```diff
@@ -0,0 +1,493 @@
+        // littering the repo with things that shouldn't be at internal visibility. As a practical
+        // matter the worst possible outcome here is some excess retain/releases, none of the
+        // code is generic, so this is basically acceptable.
+        return chain.leaf.hasValidIdentityForService(
+            serverHostname: self.convertedServerHostname,
+            serverIP: targetIP
+        )
+    }
+}
+
+extension ServerIdentityPolicy {
+    @usableFromInline
+    enum IPAddress {
+        case v4(in_addr)
+        case v6(in6_addr)
+    }
+
+    @usableFromInline
+    enum LazyIPAddress {
+        case ipAddress(IPAddress)
+        case string(String)
+    }
+
+    @usableFromInline
+    struct PreparedServerHostname {
+        var bytes: ArraySlice<UInt8>
+        var firstPeriodIndex: ArraySlice<UInt8>.Index?
+
+        /// Creates an `PreparedServerHostname`
+        ///
+        /// This consists of a non-NULL-terminated sequence of ASCII bytes and the index of the
+        /// first period in that hostname.
+        ///
+        /// If the string this method is called on contains non-ACSII code points, this constructor fails.
+        ///
+        /// This constructor exists to avoid doing repeated loops over the string buffer.
+        /// In a naive implementation we'd loop at least four times: once to lowercase
+        /// the string, once to get a buffer pointer to a contiguous buffer, once
+        /// to confirm the string is ASCII, and once to find the first period for matching wildcards.
+        /// Here we can do that all in one loop.
+        @usableFromInline
+        init?(lowercaseASCIIBytes string: String) {
+            let utf8View = string.utf8
+            self.firstPeriodIndex = nil
+
+            var value: [UInt8] = []
+            value.reserveCapacity(utf8View.count)
+
+            for codeUnit in utf8View {
+                guard codeUnit.isValidDNSCharacter else {
+                    return nil
+                }
+
+                if self.firstPeriodIndex == nil && codeUnit == asciiPeriod {
+                    // This is tricky and not generically correct, but it's safe for Array.
+                    self.firstPeriodIndex = value.endIndex
+                }
+
+                // We know we have only ASCII printables, we can safely unconditionally set the 6 bit to 1 to lowercase.
+                value.append(codeUnit | (0x20))
+            }
+
+            self.bytes = value[...]
+
+            // Strip trailing period.
+            if self.bytes.last == asciiPeriod {
+                self.bytes = self.bytes.dropLast()
+            }
+        }
+    }
+}
+
+extension Optional where Wrapped == ServerIdentityPolicy.LazyIPAddress {
+    /// Converts the value from string to one of the IP address cases and
+    /// returns it.
+    ///
+    /// Does nothing if the value is nil.
+    @inlinable
+    mutating func convert() -> ServerIdentityPolicy.IPAddress? {
```

Comment:
NIT: Shouldn't this rather be an init on `ServerIdentityPolicy.IPAdress` that takes a `ServerIdentityPolicy.LazyIPAddress` and on the optional we can just call `flatMap`?

Reply:
No, because we modify this value as well.